File: Pertemuan1/Mahasiswa.java
/**
 * Sesi 2 — enkapsulasi yang menjaga invariant.
 *
 * Deskripsi masalah:
 *   "Sistem akademik mencatat mahasiswa dengan NIM, nama, dan tiga komponen
 *    nilai: tugas, UTS, dan UAS. NIM tidak pernah berubah setelah mahasiswa
 *    terdaftar. Setiap komponen nilai berada dalam rentang 0 sampai 100.
 *    Nilai akhir dihitung 30% tugas, 30% UTS, 40% UAS."
 *
 * Tuliskan lebih dulu daftar invariant-nya di analisis.md,
 * baru lengkapi TODO di bawah ini.
 */
public class Mahasiswa {

    // Konstanta bobot — jangan menulis angka 0.30 dan 0.40 di dalam method.
    public static final double BOBOT_TUGAS = 0.30;
    public static final double BOBOT_UTS   = 0.30;
    public static final double BOBOT_UAS   = 0.40;

    private static final double NILAI_MIN = 0;
    private static final double NILAI_MAX = 100;

    // TODO 1: deklarasikan atribut. Perhatikan mana yang boleh berubah
    //         dan mana yang tidak. Gunakan final untuk yang tidak boleh berubah.
    private final String nim;
    private final String nama;
    private double nilaiTugas;
    private double nilaiUts;
    private double nilaiUas;

    public Mahasiswa(String nim, String nama, double nilaiTugas, double nilaiUts, double nilaiUas) {
        // TODO 2: tolak NIM yang kosong atau null.
        //         Lemparkan IllegalArgumentException dengan pesan yang menyebut
        //         APA yang salah — bukan sekadar "Error".
        if (nim == null || nim.isEmpty()) {
            throw new IllegalArgumentException("Nim tidak boleh kosong");
        }
        // TODO 3: tolak setiap komponen nilai yang di luar rentang 0-100.
        //         Petunjuk: buat satu method privat pembantu agar tidak menulis
        //         pemeriksaan yang sama tiga kali.
        if (nilaiTugas < NILAI_MIN || nilaiTugas > NILAI_MAX) {
            throw new IllegalArgumentException("Nilai tugas tidak boleh di bawah " + NILAI_MIN + " mahi di atas " + NILAI_MAX);
            }
        
        this.nim = nim;
        this.nama = nama;
        this.nilaiTugas = nilaiTugas;
        this.nilaiUts = nilaiUts;
        this.nilaiUas = nilaiUas;
    }

    // TODO 4: buat method privat pembantu untuk memvalidasi satu komponen nilai.
    //         Tanda tangan yang disarankan:
    //         private static void pastikanNilaiSah(String namaKomponen, double nilai)
    private static void pastikanNilaiSah(String namaKomponen, double nilai) throws IllegalArgumentException
    {
        if (nilai < NILAI_MIN || nilai > NILAI_MAX)
            throw new IllegalArgumentException(namaKomponen + " tidak boleh di bawah " + NILAI_MIN + " mahi di atas " + NILAI_MAX);
    }

    /**
     * TODO 5: hitung nilai akhir memakai konstanta bobot di atas.
     */
    public double nilaiAkhir() {
        return (nilaiTugas * BOBOT_TUGAS)
                + (nilaiUts * BOBOT_UTS)
                + (nilaiUas * BOBOT_UAS);   
    }

    /**
     * TODO 6: kembalikan huruf mutu berdasarkan nilai akhir.
     *   >= 80 -> "A"   >= 70 -> "B"   >= 60 -> "C"   >= 50 -> "D"   selain itu "E"
     */
    public String hurufMutu() {
        double nilai = nilaiAkhir();
        
        if (nilai >= 80) {
            return "A";
        } else if (nilai >= 70) {
            return "B";
        } else if (nilai >= 60) {
            return "C";
        } else if (nilai >= 50) {
            return "D";
        } else {
            return "E";
        }  
    }

    // ── Getter ────────────────────────────────────────────────
    // TODO 7: sediakan getter untuk nim, nama, dan nilaiAkhir.
    //         JANGAN membuat setNim(). Baca ulang invariant Anda kalau tergoda.
    public String getNim() {
        return nim;
    }
    public String getNama() {
        return nama;
    }
    public double getNilaiAkhir() {
        return nilaiAkhir();
    }

    @Override
    public String toString() {
        return String.format("%-10s %-18s akhir=%6.2f  mutu=%s",
                nim, nama, nilaiAkhir(), hurufMutu());
    }
}

File: Pertemuan1/Mahasiswa.php
<?php
declare(strict_types=1);

/**
 * Sesi 2 — enkapsulasi yang menjaga invariant (PHP).
 * Bandingkan baris demi baris dengan java/Mahasiswa.java.
 */
class Mahasiswa
{
    public const float BOBOT_TUGAS = 0.30;
    public const float BOBOT_UTS   = 0.30;
    public const float BOBOT_UAS   = 0.40;

    private const float NILAI_MIN = 0;
    private const float NILAI_MAX = 100;

    /**
     * Constructor property promotion (PHP 8):
     * readonly adalah padanan `final` pada atribut Java.
     *
     * TODO 1: lengkapi daftar parameter — tentukan mana yang readonly.
     */
        public function __construct(
        private readonly string $nim,
        private readonly string $nama,
        private float $nilaiTugas,
        private float $nilaiUts,
        private float $nilaiUas,
    ) {
        // TODO 2: tolak NIM yang kosong (setelah di-trim).
        if (trim($nim) === '') {
            throw new InvalidArgumentException('NIM tidak boleh kosong.');
        }

        // TODO 3: tolak setiap komponen nilai di luar rentang 0-100
        self::pastikanNilaiSah('nilaiTugas', $nilaiTugas);
        self::pastikanNilaiSah('nilaiUts', $nilaiUts);
        self::pastikanNilaiSah('nilaiUas', $nilaiUas);
    }

    /**
     * TODO 4: lengkapi validasi satu komponen nilai.
     */
    private static function pastikanNilaiSah(string $namaKomponen, float $nilai): void
    {
        if ($nilai < self::NILAI_MIN || $nilai > self::NILAI_MAX) {
            throw new InvalidArgumentException(sprintf(
                '%s harus berada pada rentang %.0f-%.0f, diterima: %.2f',
                $namaKomponen, self::NILAI_MIN, self::NILAI_MAX, $nilai
            ));
        }
    }

    /** TODO 5: hitung nilai akhir memakai konstanta bobot. */
    public function nilaiAkhir(): float
    {
        return $this->nilaiTugas * self::BOBOT_TUGAS
             + $this->nilaiUts   * self::BOBOT_UTS
             + $this->nilaiUas   * self::BOBOT_UAS;
    }

    /** TODO 6: kembalikan huruf mutu. Petunjuk: match (true) { ... } */
    public function hurufMutu(): string
    {
        $akhir = $this->nilaiAkhir();

        return match (true) {
            $akhir >= 85 => 'A',
            $akhir >= 75 => 'B',
            $akhir >= 60 => 'C',
            $akhir >= 45 => 'D',
            default      => 'E',
        };
    }

    // TODO 7: sediakan getter seperlunya. JANGAN membuat setNim().
    public function getNim(): string  { return $this->nim; }
    public function getNama(): string { return $this->nama; }
    public function getNilaiTugas(): float { return $this->nilaiTugas; }
    public function getNilaiUts(): float   { return $this->nilaiUts; }
    public function getNilaiUas(): float   { return $this->nilaiUas; }

    public function __toString(): string
    {
        return sprintf('%-10s %-18s akhir=%6.2f  mutu=%s',
            $this->nim, $this->nama, $this->nilaiAkhir(), $this->hurufMutu());
    }
}